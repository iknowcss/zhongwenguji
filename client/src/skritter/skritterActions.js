import getConfig from '../getConfig';
import { isLoggedIn } from './skritterReducer';

export const actionTypes = {
  CONTEXT_FETCH_START: '@zwgj//skritterActions/context/fetch/start',
  CONTEXT_FETCH_FAIL: '@zwgj//skritterActions/context/fetch/fail',
  CONTEXT_FETCH_SUCCESS: '@zwgj//skritterActions/context/fetch/success',
  ADD_START: '@zwgj//skritterActions/add/start',
  LOGIN_START: '@zwgj//skritterActions/login/start'
};

export const addToSkritter = () => (dispatch, getState) => {
  if (!isLoggedIn(getState())) {
    dispatch({ type: actionTypes.LOGIN_START });
    const reduxState = JSON.stringify(getState());
    localStorage.setItem('reduxState', reduxState);
    window.location.assign(getConfig().skritterCallbackUrl);
  } else {
    dispatch({ type: actionTypes.ADD_START });
  }
};

export const fetchContext = code => (dispatch) => {
  dispatch({ type: actionTypes.CONTEXT_FETCH_START });

  return fetch(getConfig().skritterContextUrl, {
    mode: 'cors',
    cache: 'no-cache',
    credentials: 'omit',
    redirect: 'follow',
    headers: { Authorization: `Code ${code}` }
  })
    .then((resp) => {
      const { status, statusText } = resp;
      if (status !== 200) {
        return Promise.reject(new Error(`Unexpected HTTP response: ${status} ${statusText}`));
      }
      return resp.json();
    })
    .then((context) => {
      dispatch({
        type: actionTypes.CONTEXT_FETCH_SUCCESS,
        context
      });
    })
    .catch((error) => {
      console.error('Failed to fetch skritter context', error);
      dispatch({ type: actionTypes.CONTEXT_FETCH_FAIL });
    });
};

