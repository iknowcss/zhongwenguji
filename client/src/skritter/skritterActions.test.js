import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockConsole from 'jest-mock-console';
import rootReducer from '../reducer';
import { actionTypes, addToSkritter, fetchContext } from './skritterActions';

jest.mock('../getConfig', () => () => ({
  skritterCallbackUrl: 'http://example.com/authorize',
  skritterContextUrl: 'http://example.com/skritter/context'
}));

describe('skritterActionCreator', () => {
  const buildMockStore = configureMockStore([thunk]);
  let store;

  describe('addToSkritter', () => {
    let initialState;

    describe('not logged in', () => {
      beforeAll(() => {
        window.location.assign = jest.fn();
        initialState = rootReducer();
        store = buildMockStore(initialState);
        store.dispatch(addToSkritter());
      });

      it('stores the current redux state to the session storage', () => {
        expect(JSON.parse(localStorage.__STORE__['reduxState']))
          .toEqual(initialState);
      });

      it('redirects to the skritter OAuth2 endpoint', () => {
        expect(store.getActions()[0]).toEqual({
          type: actionTypes.LOGIN_START
        });
        expect(window.location.assign)
          .toHaveBeenCalledWith('http://example.com/authorize');
      });
    });

    describe('logged in', () => {
      beforeAll(() => {
        window.location.assign = jest.fn();
        initialState = rootReducer(rootReducer(), {
          type: actionTypes.CONTEXT_FETCH_SUCCESS,
          context: {
            auth: 'b2hhaQ==',
            user: { name: 'iknowcss' }
          }
        });
        store = buildMockStore(initialState);
        store.dispatch(addToSkritter());
      });

      it('starts adding', () => {
        expect(store.getActions()).toEqual([{ type: actionTypes.ADD_START }]);
      });
    });
  });

  describe('fetchContext', () => {
    let restoreConsole;

    beforeEach(() => {
      fetch.resetMocks();
      store = buildMockStore();
      restoreConsole = mockConsole();
    });

    afterEach(() => {
      restoreConsole();
    });

    it('fetches the skritter from the server', async () => {
      fetch.mockResponseOnce(JSON.stringify({ foo: 'bar' }));

      await expect(store.dispatch(fetchContext('secretcode')))
        .resolves.toBeUndefined();

      expect(fetch.mock.calls).toHaveLength(1);
      expect(fetch.mock.calls[0]).toMatchSnapshot();

      const actions = store.getActions();
      expect(actions[0]).toEqual({ type: actionTypes.CONTEXT_FETCH_START });
      expect(actions[1]).toEqual({
        type: actionTypes.CONTEXT_FETCH_SUCCESS,
        context: { foo: 'bar' }
      });
    });

    it('handles unexpected response codes', async () => {
      fetch.mockResponseOnce('', { status: 401, statusText: 'Not authorized' });

      await expect(store.dispatch(fetchContext('secretcode')))
        .resolves.toBeUndefined();

      const actions = store.getActions();
      expect(actions[0]).toEqual({ type: actionTypes.CONTEXT_FETCH_START });
      expect(actions[1]).toEqual({ type: actionTypes.CONTEXT_FETCH_FAIL });

      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error.mock.calls[0]).toEqual([
        'Failed to fetch skritter context',
        expect.objectContaining({
          message: 'Unexpected HTTP response: 401 Not authorized'
        })
      ]);
    });

    it('handles network errors', async () => {
      fetch.mockRejectOnce(new Error('Failed to connect'));

      await expect(store.dispatch(fetchContext('secretcode')))
        .resolves.toBeUndefined();

      const actions = store.getActions();
      expect(actions[0]).toEqual({ type: actionTypes.CONTEXT_FETCH_START });
      expect(actions[1]).toEqual({ type: actionTypes.CONTEXT_FETCH_FAIL });

      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error.mock.calls[0]).toEqual([
        'Failed to fetch skritter context',
        expect.objectContaining({
          message: 'Failed to connect'
        })
      ]);
    });
  });
});
