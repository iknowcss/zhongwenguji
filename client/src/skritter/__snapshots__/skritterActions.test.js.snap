// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skritterActionCreator fetchContext success fetches the skritter context from the server 1`] = `
Array [
  "http://example.com/skritter/context",
  Object {
    "cache": "no-cache",
    "credentials": "omit",
    "headers": Object {
      "Authorization": "Code secretcode",
    },
    "mode": "cors",
    "redirect": "follow",
  },
]
`;

exports[`skritterActionCreator submitToSkritter success submits the vocab list 1`] = `
Object {
  "cache": "no-cache",
  "credentials": "omit",
  "headers": Object {
    "content-type": "application/json",
    "x-session": "mock-session",
  },
  "method": "POST",
  "mode": "cors",
  "redirect": "follow",
}
`;
